packages/firestore/src/model/path.ts

```diff
@@ -63,31 +46,34 @@ abstract class Path {
   }
 
   /**
-   * Constructs a new instance of Path using the same concrete type as `this`.
-   * We need this instead of using the normal constructor, because polymorphic
-   * `this` doesn't work on static methods.
+   * Abstract constructor method to construct an instance of B with the given
+   * parameters.
    */
-  private construct(
+  protected abstract construct(
     segments: string[],
     offset?: number,
     length?: number
-  ): this {
-    const path: this = Object.create(Object.getPrototypeOf(this));
-    path.init(segments, offset, length);
-    return path;
-  }
+  ): B;
+
+  /**
+   * Returns a String representation.
+   *
+   * Implementing classes are required to provide deterministic implementations as
+   * the String representation is used to obtain canonical Query IDs.
+   */
+  abstract toString(): string;
 
   get length(): number {
     return this.len;
   }
 
-  isEqual(other: Path): boolean {
-    return Path.comparator(this, other) === 0;
+  isEqual(other: this): boolean {
```

Comment:
Should this now also be `B`?

Reply:
Sure.  That makes it so you can't use .isEqual() to check for equality between a `ResourcePath` and a `FieldPath` but that's probably good. 👍